go/oasis-test-runner/scenario/e2e/halt_restore.go

```diff
@@ -0,0 +1,162 @@
+package e2e
+
+import (
+	"context"
+	"fmt"
+	"path/filepath"
+	"time"
+
+	"github.com/oasislabs/oasis-core/go/common/logging"
+	"github.com/oasislabs/oasis-core/go/oasis-test-runner/env"
+	"github.com/oasislabs/oasis-core/go/oasis-test-runner/oasis"
+	"github.com/oasislabs/oasis-core/go/oasis-test-runner/scenario"
+)
+
+var (
+	// HaltRestore is the halt and restore scenario.
+	HaltRestore scenario.Scenario = newHaltRestoreImpl()
+
+	dumpGlob = "genesis-oasis-test-runner-at-*.json"
+)
+
+type haltRestoreImpl struct {
+	basicImpl
+
+	logger *logging.Logger
+}
+
+func newHaltRestoreImpl() scenario.Scenario {
+	sc := &haltRestoreImpl{
+		basicImpl: basicImpl{
+			clientBinary: "test-long-term-client",
+			clientArgs:   []string{"--mode", "part1"},
+		},
+		logger: logging.GetLogger("scenario/e2e/halt_restore"),
+	}
+	return sc
+}
+
+func (sc *haltRestoreImpl) Fixture() (*oasis.NetworkFixture, error) {
+	f, err := sc.basicImpl.Fixture()
+	if err != nil {
+		return nil, err
+	}
+	f.Network.HaltEpoch = 2
+	f.Network.EpochtimeMock = true
+	return f, nil
+}
+
+func (sc *haltRestoreImpl) Name() string {
+	return "halt-restore"
+}
+
+func (sc *haltRestoreImpl) Run(childEnv *env.Env) error {
+	clientErrCh, cmd, err := sc.basicImpl.start(childEnv)
+	if err != nil {
+		return err
+	}
+
+	ctx := context.Background()
+	logger.Info("waiting for nodes to register")
+	if err = sc.net.Controller().WaitNodesRegistered(ctx, sc.net.NumRegisterNodes()); err != nil {
+		return fmt.Errorf("scenario/e2e/halt_restore: waiting for nodes to register: %w", err)
+	}
+	logger.Info("nodes registered")
+
+	logger.Info("triggering epoch transition")
+	if err = sc.net.Controller().SetEpoch(ctx, 1); err != nil {
+		return fmt.Errorf("scenario/e2e/halt_restore: setting epoch: %w", err)
+	}
+	logger.Info("epoch transition done")
+
+	// Wait for the client to exit.
+	select {
+	case err = <-sc.basicImpl.net.Errors():
+		_ = cmd.Process.Kill()
+	case err = <-clientErrCh:
+	}
+	if err != nil {
+		return err
+	}
+
+	logger.Info("triggering epoch transition to halt epoch")
+	if err = sc.net.Controller().SetEpoch(ctx, 2); err != nil {
+		return fmt.Errorf("scenario/e2e/halt_restore: setting halt epoch: %w", err)
+	}
+	logger.Info("halt epoch transition done")
+
+	// TODO: add a 'wait-epoch' command, switch to epochtime backend, and wait for 3rd epoch here.
```

Comment:
Either a command or a method on the `sc.net.Controller()`?

But wait, if this is the mock epochtime backend, there will be no 3rd epoch unless you call `SetEpoch` (which already waits for the epoch transition to happen)?

Reply:
yeah with a method on the `sc.net.Controller()` i'd switch to the non-mock backend

Reply:
But why not just keep the mock backend and call `SetEpoch`? It should behave the same as if you wait for the epoch transition in a non-mock backend.

The only problem currently is that this will cause the nodes to terminate, so you need to make sure this is allowed in this test (also see #2296 which adds support for allowing node shutdown in tests).

Reply:
once we're in the "halt" epoch we aren't able to switch epochs anymore with the mock backend (since consensus is in read-only mode). However with the non mock backend the epoch will change since blocks continue to happen.

Reply:
Oh, oops, yeah that's true :-) So you'll add the RPC method in this PR? Could be part of the client service (`go/client`).

Reply:
yeah i'll do it :+1:

Reply:
Updated the test, using the normal epochtime now. 

Cannot actually wait for the epoch after the halt epoch though, as the block in which the consensus halts is never produced (which is the correct behavior) - so I currently ignore the the error in the test and it actually works more like a "wait for shutdown" command. See the comment in the code.

Also the panic in the consensus doesn't actually shutdown the node at the moment (i think this is fixed in https://github.com/oasislabs/oasis-core/pull/2296) so updating the test to allow node shutdowns should be done after either this or #2296 PR is merged.

Reply:
yeah should just be changed to wait for node termination (similarly as it's done in #2296) when these branches are merged